metronome/checkpointing/service/src/io/iohk/metronome/checkpointing/service/CheckpointingService.scala

```diff
@@ -0,0 +1,239 @@
+import cats.effect.concurrent.Ref
+import cats.effect.{Concurrent, Resource, Sync}
+import cats.implicits._
+import io.iohk.metronome.checkpointing.CheckpointingAgreement
+import io.iohk.metronome.checkpointing.models.{
+  Block,
+  CheckpointCertificate,
+  Ledger
+}
+import io.iohk.metronome.checkpointing.service.CheckpointingService.LedgerTree
+import io.iohk.metronome.checkpointing.service.storage.LedgerStorage
+import io.iohk.metronome.crypto.ECPublicKey
+import io.iohk.metronome.hotstuff.consensus.basic.{Phase, QuorumCertificate}
+import io.iohk.metronome.hotstuff.service.ApplicationService
+import io.iohk.metronome.hotstuff.service.storage.{
+  BlockStorage,
+  ViewStateStorage
+}
+import io.iohk.metronome.storage.KVStoreRunner
+
+import scala.annotation.tailrec
+
+class CheckpointingService[F[_]: Sync, N](
+    ledgerTree: Ref[F, LedgerTree],
+    lastExecutedHeader: Ref[F, Block.Header],
+    ledgerStorage: LedgerStorage[N],
+    blockStorage: BlockStorage[N, CheckpointingAgreement]
+)(implicit storeRunner: KVStoreRunner[F, N])
+    extends ApplicationService[F, CheckpointingAgreement] {
+
+  override def createBlock(
+      highQC: QuorumCertificate[CheckpointingAgreement]
+  ): F[Option[Block]] = ???
+
+  override def validateBlock(block: Block): F[Option[Boolean]] = {
+    val ledgers = for {
+      nextLedger <- OptionT(projectLedger(block))
+      tree       <- OptionT.liftF(ledgerTree.get)
+      prevLedger <- tree.get(block.header.parentHash).map(_._1).toOptionT[F]
+    } yield (prevLedger, nextLedger)
+
+    ledgers.value.flatMap {
+      case Some((prevLedger, nextLedger))
+          if nextLedger.hash == block.header.postStateHash =>
+        validateTransactions(block.body, prevLedger)
+
+      case _ => false.some.pure[F]
+    }
+  }
+
+  private def validateTransactions(
+      body: Block.Body,
+      ledger: Ledger
+  ): F[Option[Boolean]] = {
+    //TODO: Validate transactions PM-3131/3132
+    true.some.pure[F]
+  }
+
+  override def executeBlock(
+      block: Block,
+      commitQC: QuorumCertificate[CheckpointingAgreement],
+      commitPath: NonEmptyList[Block.Hash]
+  ): F[Boolean] = {
+    require(commitQC.phase == Phase.Commit, "Commit QC required")
+    projectLedger(block).flatMap {
+      case Some(ledger) =>
+        if (block.hash != commitQC.blockHash)
+          saveLedger(block.header, ledger).as(true)
+        else
+          for {
+            //TODO: PM-3110:
+            // chkpOpt <- constructCheckpoint(ledger, commitQC)
+            _ <- saveLedger(block.header, ledger)
+            //TODO: PM-3110:
+            // _ <- chkpOpt.map(pushCheckpoint).getOrElse(().pure[F])
+          } yield true
+
+      case None =>
+        Sync[F].raiseError(
+          new IllegalStateException(s"Could not execute block: ${block.hash}")
+        )
+    }
+  }
+
+  private def projectLedger(block: Block): F[Option[Ledger]] = {
+    (for {
+      ledgers    <- ledgerTree.get
```

Comment:
To avoid having to choose a non-clashing name for the thing I `.get` from a `Ref`, I adopted a postfix naming convention, so the class level variable would be `ledgerTreeRef: Ref[F, LedgerTree]` and the variable just `ledgerTree`. Not sure it bothered you, just thought I'd mention it.

Reply:
Cool, I'll add this to #54